text/0000-pub-macro-rules.md

```diff
@@ -0,0 +1,300 @@
+The following are how common patterns in macros today translate to the next path based scoping system.
+
+### Deeply nested macros
+
+Macro use makes all macros inside a child module available to the parent module.
+
+```rust 
+
+#[macro_use]
+mod m {
+    #[macro_use]
+    mod n {
+        macro_rules! define_foo {  () => { fn foo() {} } }
+    }
+}
+    
+    
+define_foo!();
+```
+
+This would be translated as: 
+
+```rust 
+pub(crate) mod m {
+    pub(crate) mod n {
+        pub(crate) macro_rules! define_foo {  () => { fn foo() {} } }
+    }
+}
+    
+    
+m::n::define_foo!();
+```
+
+### Recursive macros
+
+Recursive macros are macros that call themselves (perhaps with different arguments)
+
+```rust 
+#[macro_use]
+mod m {
+    macro_rules! print_expr {
+        ($e:expr) => {{
+            println!("Going to do {}", stringify!($e));
+            print_expr!(no_print => $e)
+        }};
+        (no_print => $e:expr) => {{
+            $e
+        }};
+    }
+}
+
+fn main() {
+    print_expr!(1 + 1)
+}
+```
+
+Naively changing this to path based scope would not work as it is not guaranteed that the unqualified `print_expr` name is in scope. In the example above, `print_expr!` is used recursively inside the macro, but in a path scoped system the recursive call would not be in scope if the macro was called with a qualified path (e.g., user calls `m::print_expr!` which references unqualified `print_expr!` which is not in scope).
+
+A possible way to handle this is to introduce a new macro specific keyword `$self` which is directly analogous to `$crate` except that it refers to the module where the macro is defined. This would work in the simple case but quickly breaks down in more complicated module paths. For example:
+
+```rust 
+pub mod m {
+    mod n { // n is private
+        macro_rules! my_macro {
+            () => {
+                $self::my_macro!(@)
+            };
+            (@) => {};
+        }
+    }
+    pub use n::my_macro;
+}
+m::my_macro!()
+```
+
+If `$self` refers to the module where `my_macro!` is defined (i.e., `m::n`) then the call in the top level module would expand to `m::n::my_macro!(@)` and an error would occur due to `n` being private.
+
+`$self` could refer to the current namespace of the top-level macro call, but we are unsure at this time if this would lead to ambiguities.
+
+### "Private" macros
+
+Macros can use "private" macros (i.e., macros defined inside of other macros). This can lead to an issue where a macro is defined twice which would lead to name clashing.
+
+```rust 
+macro_rules! private {
+    () => {
+        private!(@);
+        private!(@);
+    };
+    (@) => {
+        macro_rules! __private {
+            () => {};
+        }
+    }
+}
+
+private!()
+```
+
+Here `__private` is defined twice which leads to an error.
+
+How this should be overcome is not yet known. Possible ideas are:
+* allowing macros defined inside of other macros to shadow one another. This may lead to ambiguities.
```

Comment:
```suggestion
* allowing macros defined inside of other macros to shadow one another (possibly through some opt-in attribute annotation; see _Opting into textual scoping_ below)). This may lead to ambiguities.
```